### PR TITLE
Better process broken parts on table start for replicated tables

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 20b895487228a16d06899f3453b1accdec563322 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56142/20b895487228a16d06899f3453b1accdec563322/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6699350990">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56142/20b895487228a16d06899f3453b1accdec563322/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/MergeTree/MergeTreePartInfo.cpp [hunk 1]
```diff
@@ -300,6 +300,33 @@ void MergeTreePartInfo::deserialize(ReadBuffer & in)
...
 
+bool MergeTreePartInfo::areAllBlockNumbersCovered(const MergeTreePartInfo & blocks_range, std::vector<MergeTreePartInfo> candidates)
+{
+    if (candidates.empty())
+        return false;
+
+    std::sort(candidates.begin(), candidates.end());
+
+    /// First doesn't cover left border
+    if (candidates[0].min_block > blocks_range.min_block)
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreePartInfo.cpp:311]:
```suggestion
    if (candidates[0].min_block != blocks_range.min_block)
```

File: src/Storages/MergeTree/MergeTreePartInfo.cpp [hunk 2]
```diff
@@ -300,6 +300,33 @@ void MergeTreePartInfo::deserialize(ReadBuffer & in)
...
+        return false;
+
+    int64_t current_right_block = candidates[0].min_block;
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreePartInfo.cpp:314]:
```suggestion
    int64_t current_right_block = candidates[0].min_block - 1;
```

File: src/Storages/MergeTree/MergeTreePartInfo.cpp [hunk 3]
```diff
@@ -300,6 +300,33 @@ void MergeTreePartInfo::deserialize(ReadBuffer & in)
...
+
+    for (const auto & candidate : candidates)
+    {
+        if (candidate.min_block - current_right_block > 1)
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreePartInfo.cpp:318]:
```suggestion
        if (current_right_block + 1 != candidate.min_block)
```

File: src/Storages/MergeTree/MergeTreePartInfo.cpp [hunk 4]
```diff
@@ -300,6 +300,33 @@ void MergeTreePartInfo::deserialize(ReadBuffer & in)
...
+            return false;
+
+        current_right_block = candidate.max_block;
+    }
+
+    if (current_right_block < blocks_range.max_block)
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreePartInfo.cpp:324]:
```suggestion
    if (current_right_block != blocks_range.max_block)
```

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Integration tests (asan) [2/6] - https://github.com/ClickHouse/ClickHouse/issues/55090
Integration tests (release) [2/4] - https://github.com/ClickHouse/ClickHouse/issues/52554

**Pull request merged**